File: src/shared/api/notes.ts
import { invoke } from '@tauri-apps/api/core';
import { Note, TextRange } from '../types/note';



// API functions
export async function addNote(
	bookPath: string,
	range: TextRange,
	text: string,
	preview: string,
	highlight: boolean = false,
	highlightColor: string | null = null,
): Promise<Note> {
	return invoke<Note>('add_note', {
		bookPath,
		range,
		text,
    preview,
		highlight,
		highlightColor,
	});
}

export async function getNotes(bookPath?: string): Promise<Note[]> {
  return invoke<Note[]>('get_notes', { bookPath: bookPath ?? null });
}

export async function updateNote(
	noteId: string,
	range: TextRange,
	text: string,
	highlight: boolean,
	highlightColor?: string,
): Promise<Note> {
	return invoke<Note>('update_note', {
		noteId,
		range,
		text,
		highlight,
		highlightColor,
	});
}

export async function deleteNote(noteId: string): Promise<void> {
  return invoke('delete_note', { noteId });
}

File: src-tauri/src/core/reader/notes.rs
use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::core::reader::TextLocation;

#[derive(Serialize, Deserialize, Debug, Clone, Default)]
pub struct TextRange {
    pub start: TextLocation,
    pub end: TextLocation,
}

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct Note {
    pub id: String,
    pub preview: String,
    pub book_path: String,
    pub range: TextRange,
    pub text: String,
    pub highlight: bool,
    pub highlight_color: Option<String>,
    pub created_at: i64,
    pub updated_at: i64,
}

impl Note {
    pub fn new(
        book_path: String,
        range: TextRange,
        text: String,
        preview: String,
        highlight: bool,
        highlight_color: Option<String>,
        created_at: i64,
    ) -> Self {
        Self {
            id: Uuid::new_v4().to_string(),
            book_path,
            range,
            text,
            preview,
            highlight,
            highlight_color,
            created_at,
            updated_at: created_at,
        }
    }
}
